### DISCUSSION TITLE
function normals() produces errors when validating the glb

### DISCUSSION THREAD

**Discussion by fernandovt-lab** [donmccurdy/glTF-Transform#813] [Q&A] [answered]
Hi! I am building a GLB from scratch. I wanted to regenerate the normals using this function, since the ones that come with the geometry I used to build the GLB sometimes, the model looks incorrect with some parts being improperly shadowed. The issue is that after using this function, when I pass the GLB through a validator, it shows hundreds or thousands of errors, such as the following examples:
```
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 108..110 is not of unit length: 0. | /meshes/2/primitives/0/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 111..113 is not of unit length: 0. | /meshes/2/primitives/0/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 114..116 is not of unit length: 0. | /meshes/2/primitives/0/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 36..38 is not of unit length: 0. | /meshes/2/primitives/1/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 39..41 is not of unit length: 0. | /meshes/2/primitives/1/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 42..44 is not of unit length: 0. | /meshes/2/primitives/1/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 8163..8165 is not of unit length: 0. | /meshes/2/primitives/2/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 8166..8168 is not of unit length: 0. | /meshes/2/primitives/2/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 8169..8171 is not of unit length: 0. | /meshes/2/primitives/2/attributes/NORMAL
ACCESSOR_VECTOR3_NON_UNIT | Vector3 at accessor indices 54..56 is not of unit length: 0. | /meshes/27/primitives/2/attributes/NORMAL
```

Using the original normals does not give any errors, but it produces the error mentioned above. Not using any normals or using the normals() function improves the appearance, but the function normals() causes the model to have those errors when it is passed through a validator

I found you talking about this topic before https://github.com/donmccurdy/glTF-Transform/discussions/263 , and searched the internet for Zero-length normal vector. Found this stackoverflow question https://stackoverflow.com/questions/722073/how-do-you-normalize-a-zero-vector , witch the main answer seems like it suggest to return a 0. Might this be a possibility? Another posibility might be for the function normals(), to keep the original normals in case that position index does not produce valid normals?

**Comment by donmccurdy** [comment, marked as answer]
Hi @fernandovt-lab!

Related implementation, where `normalize()` is imported from `gl-matrix`:

https://github.com/donmccurdy/glTF-Transform/blob/55c779689ac135a2d4206dac2fdd3c09a6186169/packages/functions/src/normals.ts#L84-L94

My guess would be that (a) the model contains many degenerate triangles, or (b) you're creating mesh primitives using GL Draw modes other than triangles — lines, points, or triangle fans? Any idea which might be happening, or can you share an example model?

I think the `normals()` implementation could be improved to support indices and other draw modes, but I'd want to check if that's actually the issue first.

**Reply by fernandovt-lab** [reply]
The model does only have triangles, but its true its using indices. I am also doing some transformations using transformPrimitive()

Here are two examples models, one using the original mormals, and the other one using the normals from normals() function with 75 errors

[models.zip](https://github.com/donmccurdy/glTF-Transform/files/10705455/models.zip)

**Reply by donmccurdy** [reply]
Thanks! I tried loading the model into https://gltf.report and running the script below:

```typescript
import { vec3 } from '@gltf-transform/core';
import { normals } from '@gltf-transform/functions';
import { vec3 as gl } from 'gl-matrix';

await document.transform(normals({overwrite: true}));

//

const v0 = [0, 0, 0] as vec3;
const v1 = [0, 0, 0] as vec3;
const getArea = (a: vec3, b: vec3, c: vec3): number => {
	gl.sub(v0, c, b);
	gl.sub(v1, a, b);
	return gl.len(gl.cross(v0, v0, v1)) * 0.5;
}

//

let numDegen = 0;

const a = [0, 0, 0] as vec3;
const b = [0, 0, 0] as vec3;
const c = [0, 0, 0] as vec3;

for (const mesh of document.getRoot().listMeshes()) {
	for (const prim of mesh.listPrimitives()) {
		const position = prim.getAttribute('POSITION');
		for (let i = 0, il = position.getCount(); i < il; i += 3) {
			position.getElement(i, a);
			position.getElement(i + 1, b);
			position.getElement(i + 2, c);
			if (getArea(a, b, c) <= 0) {
				numDegen++;
			}
		}
		
	}
}

console.log(`Found ${numDegen} degenerate triangles.`);
```

It finds 25 degenerate triangles, which matches up to the 75 non-unit normals (3 vertices per triangle * 25 triangles = 75 vertices) found by the validator. The calculation is assigning 0,0,0 as normals where the triangle is collapsed to a line or a point. I think that's more or less reasonable given the input, but if you want those triangles anyway and need to force valid output, you could modify the code above slightly, to assign `0,0,1` (or something suitably arbitrary) for the normals instead.

**Reply by fernandovt-lab** [reply]
Can't test it right now, but I guess something like this would work?

```
for (let i = 0, il = position.getCount(); i < il; i += 3) {
	position.getElement(i, a);
	position.getElement(i + 1, b);
	position.getElement(i + 2, c);
	if (getArea(a, b, c) <= 0) {
		numDegen++;
		normal.setElement(i, [0, 0, 1]);
		normal.setElement(i + 1, [0, 0, 1]);
		normal.setElement(i + 2, [0, 0, 1]);

	}
}
```

**Reply by donmccurdy** [reply]
That should work, yes!

**Reply by fernandovt-lab** [reply]
Great! I will try it then, thanks for everything!
